File: internal/model/audit.go
package model

import "time"

// 审计事件来源。
const (
	SourceWeb       = "web"
	SourceAPI       = "api"
	SourceScheduler = "scheduler"
	SourceSystem    = "system"
	// SourceEmergency 是宿主机本地执行的应急脚本（f-9-06）。
	//
	// 单独一个取值而不是并进 system：要区分「这次重置是从面板点的还是从
	// 宿主机命令行敲的」。能登上面板的人可能很多，能登上宿主机的人通常很少
	// ——而这个区别正是事后追责时最要紧的那一条。
	SourceEmergency = "emergency"
)

// AuditLog 对应 audit_log 表。
//
// 这是**只增不改**的流水表：任何写操作与安全事件都应记录（DATA_MODEL §4.14）。
// 操作人名称与资源名以「快照」形式冗余存储——用户改名或资源删除后，
// 历史审计仍须能还原当时的事实。
type AuditLog struct {
	ID           int64     `gorm:"primaryKey"`
	At           time.Time `gorm:"not null"`
	OperatorID   *int64
	OperatorName *string `gorm:"size:64"`
	Source       string  `gorm:"size:16;not null;default:web"`
	NodeID       *int64
	ResourceType string `gorm:"size:32;not null"`
	ResourceID   *int64
	ResourceName *string `gorm:"size:128"`
	Action       string  `gorm:"size:48;not null"`
	Params       *string `gorm:"type:text"`
	BeforeState  *string `gorm:"type:text"`
	AfterState   *string `gorm:"type:text"`
	// Success 刻意**不带 default tag**：GORM 对带 default 的字段会跳过零值，
	// 导致失败记录（false）被数据库默认值 true 覆盖——审计会因此把失败写成成功。
	// 表结构本身仍保留默认值，模型这里只约束写入行为。
	Success  bool    `gorm:"not null"`
	Error    *string `gorm:"size:512"`
	ClientIP *string `gorm:"size:64"`
}

// TableName 固定表名。
func (AuditLog) TableName() string { return "audit_log" }

File: web/src/api/audit.ts
/**
 * 审计流水接口（F-1-12）。
 *
 * **租户只能看到自己的记录**，这条隔离在服务端做，界面拿不到别人的数据。
 * 因此界面不需要、也不应该自己去判断可见性——它只管展示拿到的结果。
 */
import { get } from './client'

export interface AuditEntry {
  id: number
  at: string

  operator_id?: number
  operator_name?: string
  /**
   * 记录来源：web 还是 API 凭证。
   *
   * 它比看起来重要——用 API Key 执行的操作**不会触发二次验证**，
   * 因此「这条记录是怎么来的」是判断风险时的第一手信息。
   */
  source: string

  node_id?: number
  resource_type: string
  resource_id?: number
  resource_name?: string
  action: string

  /** 原始 JSON 文本，未解析——保留原始形态便于比对。 */
  params?: string
  before_state?: string
  after_state?: string

  success: boolean
  error?: string
  client_ip?: string
}

export interface AuditPage {
  items: AuditEntry[]
  total: number
  page: number
  page_size: number
  /** 结果被截断——审计少几条会让「没发生过」这个结论变成错的。 */
  truncated: boolean
  note?: string
}

export interface AuditFilter {
  from?: string
  to?: string
  operator_id?: number
  resource_type?: string
  action?: string
  /**
   * 来源筛选（G-38）：web（界面）/ api（API 凭证）/ emergency（带外脚本）。
   * 不传表示不限——「从哪里发起的」是判断一条记录风险的第一手信息。
   */
  source?: string
  keyword?: string
  /** 三态：不传表示不限。 */
  success?: boolean
  page?: number
  page_size?: number
}

export interface AuditFacets {
  actions: string[]
  resource_types: string[]
}

export const auditApi = {
  list: (f: AuditFilter) =>
    get<AuditPage>('/api/v1/audit', {
      ...f,
      success: f.success === undefined ? undefined : String(f.success),
    }),

  /** 筛选项由服务端给出，避免界面里的列表与后端对不上。 */
  facets: () => get<AuditFacets>('/api/v1/audit/facets'),

  get: (id: number) => get<AuditEntry>(`/api/v1/audit/${id}`),
}

export const SOURCE_LABEL: Record<string, string> = {
  web: '界面',
  api: 'API 凭证',
  system: '系统',
}

/** 把动作名渲染成可读的中文；未登记的直接显示原值以便排查。 */
export function actionLabel(action: string): string {
  const KNOWN: Record<string, string> = {
    'user.login': '登录',
    'user.logout': '登出',
    'session.revoke': '撤销会话',
    'api_key.create': '生成 API 凭证',
    'api_key.revoke': '撤销 API 凭证',
    'vm.create': '创建虚拟机',
    'vm.delete': '删除虚拟机',
    'vm.start': '开机',
    'vm.shutdown': '关机',
    'vm.migrate': '迁移虚拟机',
    'vm.reinstall': '重装系统',
    'vm.snapshot.create': '创建快照',
    'vm.snapshot.delete': '删除快照',
    'vm.snapshot.restore': '恢复快照',
    'vm.lock.release': '解锁虚拟机',
    'node.maintenance.enter': '进入维护模式',
    'node.maintenance.exit': '退出维护模式',
    'firewall.apply': '下发防火墙',
    'firewall.rollback': '回滚防火墙',
    'port_mirror.enable': '启用端口镜像',
    'port_mirror.disable': '关闭端口镜像',
    'network.repair': '修复网络',
    'network_bridge.uplink_attach': '物理口入桥',
    'storage_file.upload': '上传文件',
    'storage_file.delete': '删除文件',
    'share.mount': '挂载目录共享',
  }
  return KNOWN[action] ?? action
}
